### ISSUE TITLE
datetime.strftime("%Y") not consistent for years < 1000

### ISSUE THREAD

**Issue by eda57068-96ad-4b33-8431-9c528f59a6a6** [python/cpython#57514]
BPO | [13305](https://bugs.python.org/issue13305)
--- | :---
Nosy | @jaraco, @abalkin, @vstinner, @ned-deily, @ezio-melotti, @florentx, @vadmium, @elenaoat, @pganssle, @Fronkan
Superseder | <li>bpo-3173: external strftime for Python?</li>
Files | <li>[issue13305_xmlrpc_patch.diff](https://bugs.python.org/file23577/issue13305_xmlrpc_patch.diff "Uploaded as text/plain at 2011-10-31.23:01:06 by @florentx")</li><li>[doc.patch](https://bugs.python.org/file41476/doc.patch "Uploaded as text/plain at 2016-01-02.16:40:36 by @alessandrocucci")</li><li>[issue13305.diff](https://bugs.python.org/file42136/issue13305.diff "Uploaded as text/plain at 2016-03-11.15:41:56 by @elenaoat")</li>

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = None
created_at = <Date 2011-10-31.20:12:37.023>
labels = ['easy', 'type-bug', '3.8', '3.9', '3.7', 'library', 'docs']
title = 'datetime.strftime("%Y") not consistent for years < 1000'
updated_at = <Date 2022-02-10.19:47:24.483>
user = 'https://github.com/florentx'
```

bugs.python.org fields:
```python
activity = <Date 2022-02-10.19:47:24.483>
actor = 'jaraco'
assignee = 'docs@python'
closed = False
closed_date = None
closer = None
components = ['Documentation', 'Library (Lib)']
creation = <Date 2011-10-31.20:12:37.023>
creator = 'flox'
dependencies = []
files = ['23577', '41476', '42136']
hgrepos = []
issue_num = 13305
keywords = ['patch', 'easy']
message_count = 30.0
messages = ['146740', '146743', '146745', '146750', '146751', '146752', '146765', '146767', '146967', '146972', '146976', '146978', '252165', '257271', '257383', '257397', '261577', '261607', '261608', '261609', '261637', '261638', '261639', '261640', '307417', '359735', '359740', '394251', '413015', '413021']
nosy_count = 13.0
nosy_names = ['jaraco', 'belopolsky', 'vstinner', 'ned.deily', 'ezio.melotti', 'flox', 'docs@python', 'python-dev', 'martin.panter', 'Elena.Oat', 'p-ganssle', 'Fronkan', 'cklarson']
pr_nums = []
priority = 'normal'
resolution = None
stage = 'needs patch'
status = 'open'
superseder = '3173'
type = 'behavior'
url = 'https://bugs.python.org/issue13305'
versions = ['Python 3.7', 'Python 3.8', 'Python 3.9']
```

</p></details>

**Comment by codethief** [comment]
Given the age of the ticket and the lack of clarity on what to do, I don't think the "easy" label still applies here, does it? :-)

**Comment by jaraco** [comment]
> > I'd like for Python to provide a portable implementation of strftime instead of just documenting that the version isn't portable.
> 
> If someone wants to do that, I suggest to first start with a project on PyPI. It sounds like a tricky project. Date, time, locales and localization are hard problems!

The [wrapper in tempora](https://github.com/jaraco/tempora/blob/d46041265f5b409e7eafbe122aa4924ef13e0782/tempora/__init__.py#L71-L143) has worked well for me for a couple of years.  It defers most of the processing to the existing C implementation and merely performs the substitution early, allowing the C implementation to handle the rest.

Would it make sense for CPython to adopt a similar technique to address the concern at hand and avoid doing a full re-implementation (with all of the bugs and performance regressions that would likely entail)?

**Comment by blhsing** [comment]
> Given the age of the ticket and the lack of clarity on what to do, I don't think the "easy" label still applies here, does it? :-)

A fix for `%Y` alone would've been "easy" by `PyOS_snprintf`-formatting the year with "%04ld", but the issue is made complex when taking into account the need to support `%G`. The issue is now resolved with PR #120820 by conditionally routing the said format specifiers to `time.strftime` and wrapping the value with `PyOS_snprintf`.

**Comment by serhiy-storchaka** [comment]
Closing as a duplicate of #120713.
